clang/tools/checked-c-convert/ConstraintBuilder.cpp

```diff
@@ -588,113 +646,137 @@ class FunctionVisitor : public RecursiveASTVisitor<FunctionVisitor> {
     constrainVarsNotEq(Var, CS.getPtr());
-    constrainVarsNotEq(Var, CS.getNTArr());
   }
 
-  // constraint helpers.
-  void constraintInBodyVariable(Expr *e, ConstAtom *target) {
+  // Constraint helpers.
+  void constraintInBodyVariable(Expr *e, ConstAtom *CAtom) {
     std::set<ConstraintVariable*> Var =
       Info.getVariable(e, Context, true);
-    constrainVarsEq(Var, target);
+    constrainVarsEq(Var, CAtom);
   }
 
-  void constraintInBodyVariable(Decl *d, ConstAtom *target) {
+  void constraintInBodyVariable(Decl *d, ConstAtom *CAtom) {
     std::set<ConstraintVariable*> Var =
       Info.getVariable(d, Context, true);
-    constrainVarsEq(Var, target);
+    constrainVarsEq(Var, CAtom);
   }
 
-  // assign the provided type (target) to all the constraint variables (CVars).
+  // Assign the provided type (target) to all the constraint variables (CVars).
   void assignType(std::set<ConstraintVariable*> &CVars,
-                  ConstAtom *target) {
+                  ConstAtom *CAtom) {
     Constraints &CS = Info.getConstraints();
-    for (const auto &C : CVars)
-      C->constrainTo(CS, target);
+    for (const auto &C : CVars) {
+      C->constrainTo(CS, CAtom);
+    }
   }
 
-  // constraint all the argument of the provided call expression to be WILD.
+  // Constraint all the argument of the provided call expression to be WILD
   void constraintAllArgumentsToWild(CallExpr *E) {
     for (const auto &A : E->arguments()) {
-      // get constraint from within the function body of the caller
+      // Get constraint from within the function body of the caller.
       std::set<ConstraintVariable*> ParameterEC =
         Info.getVariable(A, Context, true);
 
       Constraints &CS = Info.getConstraints();
-      // assign WILD to each of the constraint variables.
+      // Assign WILD to each of the constraint variables.
       assignType(ParameterEC, CS.getWild());
     }
   }
 
   void arithBinop(BinaryOperator *O) {
-    constrainInBodyExprNotPtrNotNt(O->getLHS());
-    constrainInBodyExprNotPtrNotNt(O->getRHS());
+    constrainInBodyExprNotPtr(O->getLHS());
+    constrainInBodyExprNotPtr(O->getRHS());
   }
 
-  ConstAtom* getCheckedPointerConstraint(CheckedPointerKind ptrKind) {
+  ConstAtom* getCheckedPointerConstraint(CheckedPointerKind PtrKind) {
     Constraints &CS = Info.getConstraints();
-    switch(ptrKind) {
+    switch(PtrKind) {
       case CheckedPointerKind::NtArray:
         return CS.getNTArr();
       case CheckedPointerKind::Array:
         return CS.getArr();
       case CheckedPointerKind::Ptr:
         return CS.getPtr();
       case CheckedPointerKind::Unchecked:
-        llvm_unreachable("Unchecked type inside an itype. This should be impossible.");
+        llvm_unreachable("Unchecked type inside an itype. "
+                         "This should be impossible.");
     }
     assert(false && "Invalid Pointer kind.");
   }
 
+  Expr* getNormalizedExpr(Expr *E) {
+    if (dyn_cast<ImplicitCastExpr>(E)) {
```

Comment:
What happens if there are multiple ImplicitCastExprs? Should this function recursively strip off casts?